Add a CHORD metadata object, and minimally make buffer.[ch] into C++

dstndstn opened the pull request:
This builds on top of #1108 and pulls in changes bringing it equal to the dstn/frb-bf branch.

Adds a (sketch of a) CHORD metadata struct that tracks array shapes and sizes with a simple, statically-sized struct.  Adds plumbing to copy metadata objects into the CUDA GPU world.  (To-do: copy metadata *out* of the CUDA world also!)

dstndstn committed:
add CHORD metadata class; in gpuDeviceInterface, add methods for tagging a metadata structure along with GPU buffer arrays; in cudaInputData, copy reference to an incoming buffer's metadata into the GPU realm.

dstndstn committed:
sketch fixed-size ChordMetadata struct, without fancy gsl spans

dstndstn committed:
chordMetadata: get/set array dimensions & names; data generator: set CHORD metadata values, incl one-hot.

dstndstn committed:
more metadata: in data gen, look for "F" and "T" dimensions and initialize metadata; in cudaUpchannelize, adjust some of the metadata

dstndstn committed:
gitignore: ignore emacs temp files

dstndstn committed:
Cuda: Create a pipeline state object that is passed to execute();
Gpu: create a gpu frame counter that isn't modded by the gpu buffer depth and pass to queue_commands()

dstndstn committed:
cudaBasebandBeamformer: adapt to execute() call change

dstndstn committed:
cudaBasebandBeamformer: make 'info' buffer local

dstndstn committed:
gpu: keep track of GPU buffers used by each stage and add them to graphviz outputs

dstndstn committed:
gpu stages: add the option to require a pipeline state flag to run a stage

dstndstn committed:
cudaOutputData: handle case where this stage may not have run (due to _required_flag)

dstndstn committed:
various text formatting changes

dstndstn committed:
fix a stray missing int->size_t change

dstndstn committed:
misc minor changes: move code blocks, remove unused overridden functions, documentation typos

dstndstn committed:
allow creation of GPU memory buffer views (that view a subset of a larger buffer)

dstndstn committed:
define and use a float16_t type

dstndstn committed:
testDataGen: add reuse_random option and float16 values

dstndstn committed:
update cudaBasebandBeamformer kernel

dstndstn committed:
add cuda FRB beamformer (and CPU testing implementation)

dstndstn committed:
add cuda FRB beam RE-former (aka FRB2 kernel)

dstndstn committed:
add cudaQuantize kernel (aka FRB3 kernel)

dstndstn committed:
add cuda Upchannelizers (two versions with different expected number of time samples).
Erik's later work will change how the different upchannelizers are packaged;
this version is basically just here for the MVP.

dstndstn committed:
add cudaRechunk stage

dstndstn committed:
add chordMVPSetup, a Cuda stage that sets up buffer views and other things for the CHORD MVP.

dstndstn committed:
add configuration files for the MVP and for one-hot and functional testing new Cuda stages

dstndstn committed:
cmake rules: only include kotekan_gpu once even if multiple GPU backends are being built.

dstndstn committed:
some casts to avoid float16_t / __half compile errors in docker build

dstndstn committed:
test_transpose uses visUtil.hpp which includes cuda, so make sure cuda include directories are in the compile command

dstndstn committed:
fix const-ness of CUDA build() functions

dstndstn committed:
CHORD metadata: make dimension names (fixed-max-length) strings, not chars

dstndstn committed:
fix cmakelint and iwyu complaints

dstndstn committed:
Rename buffer.[ch] to buffer.[ch]pp, fix compile errors for C++ vs C compilation

dstndstn committed:
metadata: cudaOutputData copy metadata from GPU buffer to host buffer, if available;
cudaBasebandBeamformer: transfer metadata from input to output

dstndstn committed:
cudaOutputData: oops, wrong output buffer counter

dstndstn committed:
fix compile error from converting int to float16

dstndstn committed:
Merge branch 'develop' of github.com:kotekan/kotekan into dstn/mvp-3

dstndstn committed:
fix additional merge debris

dstndstn committed:
clang-format

dstndstn committed:
buffer.cpp: NULL to nullptr for ubuntu 18.04 build

andrerenard commented on lib/core/errors.h:
```diff
@@ -132,7 +132,7 @@ void set_error_message_f(const char* format, ...);
         ERROR_F(m, ##a);                                                                           \
         set_error_message_f(m, ##a);                                                               \
-        exit_kotekan(1);                                                                           \
+        exit_kotekan(ReturnCode::FATAL_ERROR);                                     \
```
Should we also "promote" other `.c/.h` files while we are at it?   I think the buffers were the last thing that needed any pure C code.

dstndstn replied:
I filed #1127 :)

dstndstn committed:
Merge branch 'develop' of github.com:kotekan/kotekan into dstn/mvp-3

andrerenard commented on config/tests/rechunk.yaml:
```diff
@@ -0,0 +1,117 @@
+##########################################
+#
```
Looks like this made it back in.

dstndstn replied:
Re-killed

andrerenard commented on lib/cuda/cudaCommand.cpp:
```diff
@@ -369,3 +369,13 @@ cudaEvent_t cudaCommand::record_end_event(int gpu_frame_id) {
     return end_events[gpu_frame_id];
 }
+
+struct chordMetadata* cudaCommand::get_chord_metadata(struct metadataContainer* mc) {
+    if (!mc)
+        return nullptr;
+    if (strcmp(mc->parent_pool->type_name, "chordMetadata")) {
+        FATAL_ERROR("cudaCommand: expected metadata to be type \"chordMetadata\", got \"{:s}\".",
+                    mc->parent_pool->type_name);
+    }
+    return static_cast<struct chordMetadata*>(mc->metadata);
+}
```
This should live in another place.  Perhap in the metadata file itself.

dstndstn replied:
Good call, done

andrerenard commented on lib/cuda/cudaInputData.cpp:
```diff
@@ -70,11 +70,20 @@ cudaEvent_t cudaInputData::execute(cudaPipelineState& pipestate,
                                   end_events[pipestate.gpu_frame_id]);
 
+    // Copy (reference to) metadata also
+    // FIXME -- need to lock here?
```
I believe the answer here is no to needing a lock.  Since you've waited for a full frame in the last function, it already must have (or not have) a metadata associated with the frame. 

As long as you complete your metadata pass/copy before you release the frame, it should be ok.  There is a bit of an issue that metadata isn't globally locked (meaning odd things could happen with multiple consumer pipelines if doing inplace changes, but that's a rare case, and out of scope here).

dstndstn replied:
Thanks!  comment removed

andrerenard commented on lib/cuda/cudaOutputData.cpp:
```diff
@@ -95,6 +96,25 @@ cudaEvent_t cudaOutputData::execute(cudaPipelineState& pipestate,
+                      _gpu_mem, pipestate.gpu_frame_id, output_buffer->buffer_name,
+                      output_buffer_execute_id);
+                output_buffer->metadata[output_buffer_execute_id] = meta;
+                // .... where is this gonna get freed?
```
When the buffer resets a frame it decrements the reference counter, so the next buffer (or one of the following ones should release it).   As long as you haven't increased the reference counter too much.   But if you do you'll know quickly, since you'll run out of metadata frames to use and the pipeline will crash.

dstndstn replied:
Thanks, and I did check that it doesn't explode in a long run :)

andrerenard commented on lib/cuda/cudaOutputData.cpp:
```diff
@@ -103,17 +123,32 @@ void cudaOutputData::finalize_frame(int frame_id) {
     } else {
-        DEBUG("Add metadata? from input {:s}[{:d}] to {:s}[{:d}]", in_buffer->buffer_name,
-              in_buffer_id, output_buffer->buffer_name, output_buffer_id);
+        // ?? does this happen?
```
I'm confused by this.  As far as I can tell `got_metadata` will always be `false` at the start of this function call...

andrerenard replied:
I'm also not sure why it's a member variable, I think it's a control variable in just this function...

andrerenard commented on lib/cuda/cudaUpchannelize.cpp:
```diff
@@ -122,6 +123,29 @@ cudaEvent_t cudaUpchannelize::execute(cudaPipelineState& pipestate,
+        assert(meta_in->get_dimension_name(2) == "F");
+        meta_out->dim[0] /= _upchan_factor;
+        meta_out->dim[2] *= _upchan_factor;
+        INFO("cudaUpchannelize: output array shape: {:s}", meta_out->get_dimensions_string());
```
Should this be `DEBUG`?   (Above as well)

dstndstn replied:
Yes, done!

andrerenard commented on lib/gpu/gpuDeviceInterface.hpp:
```diff
@@ -87,6 +91,38 @@ class gpuDeviceInterface : public kotekan::kotekanLogging {
                                  const size_t view_len);
 
+    /**
+     * @brief Fetches the metadata (if any) attached to the given GPU
```
Should likely add the `@param` and `@return` doxygen comments, since it's likely people will be looking up what these do at some point.

dstndstn replied:
Done

andrerenard commented on lib/gpu/gpuDeviceInterface.cpp:
```diff
@@ -108,7 +111,87 @@ void gpuDeviceInterface::create_gpu_memory_array_view(const std::string& source_
+    gpu_memory[name].metadata_pointers[index] = mc;
+}
+
+void gpuDeviceInterface::release_gpu_memory_array_metadata(const std::string& name,
```
I think we need to add mutex protection to all these functions.  For allocating memory currently we always have just one thread doing the precondition and execute.   However, now some of these functions operate in finalize, which is another thread. 

And of course once we support multiple pipelines, many of the gpuDevice functions will need to become threadsafe.

dstndstn replied:
Thanks for the comment, done!

andrerenard commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -0,0 +1,127 @@
+const int CHORD_META_MAX_DIMNAME = 16;
+
+struct chordMetadata {
+    struct chimeMetadata chime;
```
We likely want to make this a public inheritance, but I don't think we want to include all of `chimeMetadata` in CHORD, so some refactoring will be necessary.

andrerenard replied:
This might be something that can become a ticket.

dstndstn replied:
Indeed ;)  https://github.com/kotekan/kotekan/issues/1128

andrerenard commented on lib/stages/testDataGen.cpp:
```diff
@@ -210,6 +222,52 @@ void testDataGen::main_thread() {
+                        }
+                    }
+
+                    INFO("Chord metadata: array shape {:s}", chordmeta->get_dimensions_string());
```
`DEBUG`?

dstndstn replied:
Done!

dstndstn committed:
reply to comments from review:
- remove a bunch of now-unnecessary "struct" statements
- add locks around the gpu_memory struct in gpuDeviceInterface

dstndstn committed:
clang-format

andrerenard commented on lib/cuda/cudaOutputData.cpp:
```diff
@@ -103,17 +123,32 @@ void cudaOutputData::finalize_frame(int frame_id) {
+                  in_buffer->buffer_name, in_buffer_id, output_buffer->buffer_name,
+                  output_buffer_id);
+            pass_metadata(in_buffer, in_buffer_id, output_buffer, output_buffer_id);
+            got_metadata = true;
```
As far as I can tell this variable, `got_metadata`, is not needed.

dstndstn replied:
Good catch, that was vestigal

andrerenard reviewed (changes requested):
Thanks for the review changes!  There is just one comment around a variable that I think should be removed.

dstndstn committed:
cudaOutputData: remove un-needed "got_metadata" variable

dstndstn committed:
Remove one million instances of the word 'struct' from Buffer and various metadata structs

andrerenard reviewed (approved):
Looks like cmake-lint wants a silly change, but otherwise this is good to merge!

dstndstn committed:
one more stray struct

dstndstn commented:
Hah, so I just finished remove approximately 1,000,000 uses of `struct Buffer` from the codebase... just compiling and checking that, then I'll add that commit to this PR, if that's ok.

dstndstn committed:
clang-format

dstndstn commented:
Okay, that's in, it compiles (ship it!).  I'll be in in about 30 minutes.
